larray/core/session.py

```diff
@@ -1392,6 +1393,155 @@ def display(k, v, is_metadata=False):
+        # feed the kwargs dict with all items declared as class attributes
+        kwargs = {}
+        for key, value in vars(self.__class__).items():
+            if not key.startswith('_'):
+                kwargs[key] = value
+
+        if meta:
+            kwargs['meta'] = meta
+
+        Session.__init__(self, **kwargs)
+        object.__setattr__(self, 'add', self._disabled)
+
+        if filepath:
+            self.load(filepath)
+
+    def __setitem__(self, key, value):
+        self._check_key_value(key, value)
+
+        # we need to keep the attribute in sync (initially to mask the class attribute)
+        object.__setattr__(self, key, value)
+        self._objects[key] = value
+
+    def __setattr__(self, key, value):
+        if key != 'meta':
+            self._check_key_value(key, value)
+
+        # update the real attribute
+        object.__setattr__(self, key, value)
+        # update self._objects
+        Session.__setattr__(self, key, value)
+
+    def _check_key_value(self, key, value):
+        cls = self.__class__
+        attr_def = getattr(cls, key, None)
+        if attr_def is None:
+            raise ValueError("The '{item}' item has not been found in the '{cls}' class declaration. "
+                             "Adding a new item after creating an instance of the '{cls}' class is not permitted."
+                             .format(item=key, cls=cls.__name__))
+        if (isinstance(value, (int, float, basestring, np.generic)) and value != attr_def) \
+                or (isinstance(value, (Axis, Group)) and not value.equals(attr_def)):
+            raise TypeError("The '{key}' item is of kind '{cls_name}' which cannot by modified."
+                            .format(key=key, cls_name=attr_def.__class__.__name__))
+        if type(value) != type(attr_def):
+            raise TypeError("Expected object of type '{attr_cls}'. Got object of type '{value_cls}'."
+                            .format(attr_cls=attr_def.__class__.__name__, value_cls=value.__class__.__name__))
+        if isinstance(attr_def, Array):
+            try:
+                attr_def.axes.check_compatible(value.axes)
+            except ValueError as e:
+                msg = str(e).replace("incompatible axes:", "Incompatible axes for array '{key}':".format(key=key))
+                raise ValueError(msg)
+        elif isinstance(value, np.ndarray) and value.shape != attr_def.shape:
+            raise ValueError("Incompatible shape for Numpy array '{key}'. "
+                             "Expected shape {attr_shape} but got {value_shape}."
+                             .format(key=key, attr_shape=attr_def.shape, value_shape=value.shape))
+
+    def copy(self):
+        instance = self.__class__()
+        for key, value in self.items():
+            instance[key] = copy(value)
+        return instance
+
+    def apply(self, func, *args, **kwargs):
+        kind = kwargs.pop('kind', Array)
+        instance = self.__class__()
```

Comment:
This should either be the code in the normal Session, so that `overridding` apply() here is not necessary, or not overriden at all (because I am unsure we actually want that behaviour for FrozenSession). I would rather have it return a normal Session to not be limited by the restrictions of the FrozenSession. If we do not do that, for example, my_frozen_session.apply(Array.sum, 'gender') would not work (because the returned axes are different).